examples/psbt.rs

```diff
@@ -0,0 +1,711 @@
+
+    let psbt = sign(psbt, keys);
+
+    let got = serialize_hex(&psbt);
+    assert_eq!(got, expected_psbt_hex);
+
+    psbt
+}
+
+/// Does the combine according to the BIP, returns the combined PSBT. Verifies against BIP174 test vector.
+fn combine(mut this: Psbt, that: Psbt) -> Psbt {
+    let expected_psbt_hex = include_str!("data/combine_psbt_hex");
+
+    this.combine(that).expect("failed to combine PSBTs");
+
+    let want_psbt = hex_psbt!(expected_psbt_hex).unwrap();
+    assert_eq!(this, want_psbt);
+
+    // TODO: Something in PSBT is not roundtripping deserialize/serialize, should it?
+    // let got = serialize_hex(&this);
+    // assert_eq!(got, expected_psbt_hex);  // TODO: This call fails.
+
+    this
+}
+
+/// Does the finalize step according to the BIP, returns the combined PSBT. Verifies against BIP174
+/// test vector.
+fn finalize(psbt: Psbt) -> Psbt {
+    let expected_psbt_hex = include_str!("data/finalize_psbt_hex");
+
+    let psbt = finalize_psbt(psbt);
+
+    let want = hex_psbt!(expected_psbt_hex).unwrap();
+    assert_eq!(psbt, want);
+
+    let got = serialize_hex(&psbt);
+    assert_eq!(got, expected_psbt_hex);
+
+    psbt
+}
+
+/// Does the transaction extractor step according to the BIP, returns the combined PSBT. Verifies
+/// against BIP174 test vector.
+fn extract_transaction(psbt: Psbt) -> Transaction {
+    let expected_tx_hex = include_str!("data/extract_tx_hex");
+
+    let tx = psbt.extract_tx();
+    
+    let got = serialize_hex(&tx);
+    assert_eq!(got, expected_tx_hex);
+
+    tx
+}
+
+/// Combines two PSBTs lexicographically according to the BIP. Verifies against BIP174 test vector.
+fn combine_lexicographically() {
+    let psbt_1_hex = include_str!("data/lex_psbt_1_hex");
+    let psbt_2_hex = include_str!("data/lex_psbt_2_hex");
+
+    let expected_psbt_hex = include_str!("data/lex_combine_psbt_hex");
+    
+    let v = Vec::from_hex(&psbt_1_hex).unwrap();
+    let mut psbt_1: Psbt = deserialize(&v).expect("failed to deserialize psbt 1");
+
+    let v = Vec::from_hex(&psbt_2_hex).unwrap();
+    let psbt_2: Psbt = deserialize(&v).expect("failed to deserialize psbt 2");
+
+    psbt_1.combine(psbt_2).expect("failed to combine PSBTs");
+
+    let got = serialize_hex(&psbt_1);
+    assert_eq!(got, expected_psbt_hex);
+}
+
+/// Signs `psbt` with `keys` if required.
+fn sign(mut psbt: Psbt, keys: &[PrivateKey]) -> Psbt {
+    let secp = SECP256K1;
+
+    let mut inputs = psbt.inputs.clone();
+
+    for (i, input) in inputs.iter_mut().enumerate() {
+        let required_pks: Vec<_> = input.bip32_derivation.keys().collect();
+        for sk in keys {
+            let pk = sk.public_key(&secp);
+            if required_pks.contains(&&pk.inner) {
+                sign_psbt(&sk, &mut psbt, i).expect("failed to sign input with key");
+            }
+        }
+    }
+
+    psbt
+}
+
+/// Signs `psbt` input at `input_index` using `sk`.
+// TODO: Consider adding this code to rust-bitcoin.
```

Comment:
Warning, this TODO is non-trivial. The same signing code appears in #940, I'm not exactly sure right at this second how to handle this code or where it should go?

Reply:
We have signing code here because we're building a test or example rather than a library that does arbitrary signing. You can find [code that serves this purpose this in bdk](https://github.com/bitcoindevkit/bdk/blob/master/src/wallet/signer.rs). That seems to be place for it to me.